##### Missionaries and Cannibals

* There are N missionaries and M cannibals on the left bank of a river, and N => M
* There is a boat on the left bank of the river
* The goal is to move all missionaries and all cannibals to the right bank
* The boat can only move if there is at least one person in it
* The boat can hold at most two people
* The number of missionaries must never be less than the number of cannibals on either the left or right bank


In [12]:
print(aStarSearch(MCProblem(10,10), bfsEvaluator())[0])

['ec', 'em', 'mb', 'dm', 'mb', 'ec', 'mb', 'dc', 'mb', 'em', 'mb', 'dm', 'mb', 'ec', 'mb', 'dc', 'mb', 'em', 'mb', 'dm', 'mb', 'ec', 'mb', 'dc', 'mb', 'em', 'mb', 'dm', 'mb', 'ec', 'mb', 'dc', 'mb', 'em', 'mb', 'dm', 'mb', 'ec', 'mb', 'dc', 'mb', 'em', 'mb', 'dm', 'mb', 'ec', 'mb', 'dc', 'mb', 'em', 'mb', 'dm', 'mb', 'ec', 'mb', 'dc', 'mb', 'em', 'mb', 'dm', 'mb', 'ec', 'mb', 'dc', 'mb', 'em', 'mb', 'dm', 'mb', 'ec', 'mb', 'dc', 'mb', 'em', 'mb', 'dm', 'dc']


In [10]:
from searchFramework import aStarSearch
s = aStarSearch(MCProblem(1000,1000), bfsEvaluator())
print(s[1])
s = aStarSearch(MCProblem(1000,1000), heuristicEvaluator())
print(s[1])

(0.9375, 27981, 13989, 6)
(1.109375, 27975, 13985, 9)


In [13]:
print(aStarSearch(MCProblem(3,3), heuristicEvaluator()))

(['ec', 'em', 'mb', 'dm', 'mb', 'ec', 'mb', 'dc', 'mb', 'em', 'mb', 'dm', 'mb', 'ec', 'mb', 'dc', 'mb', 'em', 'mb', 'dm', 'dc'], (0.015625, 65, 31, 6))


In [14]:
soln, stats = aStarSearch(MCProblem(10000, 10000), bfsEvaluator())
print(stats)

(100.265625, 279981, 139989, 6)


<pre>
print(aStarSearch(MCProblem(10,10), bfsEvaluator())[0])

['ec', 'em', 'mb', 'dm', 'mb', 'ec', 'mb', 'dc', 'mb', 'em', 'mb', 'dm', 'mb', 'ec', 'mb', 'dc', 'mb', 'em', 'mb', 'dm', 'mb', 'ec', 'mb', 'dc', 'mb', 'em', 'mb', 'dm', 'mb', 'ec', 'mb', 'dc', 'mb', 'em', 'mb', 'dm', 'mb', 'ec', 'mb', 'dc', 'mb', 'em', 'mb', 'dm', 'mb', 'ec', 'mb', 'dc', 'mb', 'em', 'mb', 'dm', 'mb', 'ec', 'mb', 'dc', 'mb', 'em', 'mb', 'dm', 'mb', 'ec', 'mb', 'dc', 'mb', 'em', 'mb', 'dm', 'mb', 'ec', 'mb', 'dc', 'mb', 'em', 'mb', 'dm', 'dc']

from searchFramework import aStarSearch
s = aStarSearch(MCProblem(1000,1000), bfsEvaluator())
print(s[1])
s = aStarSearch(MCProblem(1000,1000), heuristicEvaluator())
print(s[1])

(0.9375, 27981, 13989, 6)
(1.109375, 27975, 13985, 9)

